prefect_dbt/cloud/jobs.py

```diff
@@ -318,9 +325,167 @@ def my_flow():
             f"Triggered job run with ID {run_id} was cancelled."
         )
     elif final_run_status == DbtCloudJobRunStatus.FAILED:
-        raise DbtCloudJobRunFailed(f"Triggered job run with ID: {run_id} failed.")
+        while retry_filtered_models_attempts > 0:
+            logger.info(
+                f"Retrying job run with ID: {run_id} "
+                f"{retry_filtered_models_attempts} more times"
+            )
+            try:
+                retry_filtered_models_attempts -= 1
+                run_data = await (
+                    retry_dbt_cloud_job_run_subset_and_wait_for_completion(
+                        dbt_cloud_credentials=dbt_cloud_credentials,
+                        run_id=run_id,
+                        trigger_job_run_options=trigger_job_run_options,
+                        max_wait_seconds=max_wait_seconds,
+                        poll_frequency_seconds=poll_frequency_seconds,
+                        retry_status_filters=retry_status_filters,
+                        retry_downstream_nodes=retry_downstream_nodes,
+                    )
+                )
+                return run_data
+            except Exception:
+                pass
+        else:
+            raise DbtCloudJobRunFailed(f"Triggered job run with ID: {run_id} failed.")
     else:
         raise RuntimeError(
             f"Triggered job run with ID: {run_id} ended with unexpected"
             f"status {final_run_status.value}."
         )
+
+
+def _filter_model_names_by_status(
+    run_artifact: Dict[str, Any], retry_status_filters: List[str]
+):
+    """
+    Filters model names from model results by status.
+    """
+    model_results = run_artifact.get("results", [])
+
+    model_names = set()
+    for model_result in model_results:
+        if model_result.get("status") in retry_status_filters:
+            model_id = model_result["unique_id"]
+            model_name = model_id.split(".")[-1]
+            model_names.add(model_name)
```

Comment:
I may lack some context here, so apologize ahead of time.

A couple of things I see here:
- The `run_results.json` file contains info about a completed invocation of dbt for each node (model, test, etc.).  It seems like the intention here is to only grab the models from that `results` list and, unless I'm missing something, the eventual `model_names` set will also contain test nodes as well.
- A `unique_id` for a test can sometimes contain a hash has the last value instead of the actual identifier (see example below).  dbt would give an error with 

```json
    {
      "status": "success",
      "timing": [
        {
          "name": "compile",
          "started_at": "2022-10-13T23:02:29.935132Z",
          "completed_at": "2022-10-13T23:02:30.020219Z"
        },
        {
          "name": "execute",
          "started_at": "2022-10-13T23:02:30.037553Z",
          "completed_at": "2022-10-13T23:02:30.037565Z"
        }
      ],
      "thread_id": "Thread-36",
      "execution_time": 0.35446834564208984,
      "adapter_response": {},
      "message": null,
      "failures": null,
      "unique_id": "test.tpch.dbt_constraints_source_foreign_key_tpch_lineitem_l_orderkey__o_orderkey__source_tpch_orders_.1800ab5583"
    },
```